### PR TITLE
[refactor] Refactor client and client actor to move the code for block processing to client

### PR DESCRIPTION
**Pull request by mzhangmzz**
This PR is a pure refactoring. The context is that any processing details should be put in Client instead of ClientActor. ClientActor should just serve as a coordinator class to handle messages and check triggers and immediately pass it to Client. This is better for testing since we can't write unit test for any logic in ClientActor and also better for code readability as the logic is not scattered in two classes.

This PR only moves the part around block processing. The rest is tracked by https://github.com/near/nearcore/issues/7899

### CODE REVIEW

**Comment by matklad** [comment]
>ClientActor should just serve as a coordinator class to handle messages and check triggers and immediately pass it to Client. 

Let'ss add this to ClientActor's docstring

File: chain/client/src/client.rs [hunk 1]
```diff
@@ -807,6 +810,166 @@ impl Client {
...
+        block: Block,
+        peer_id: PeerId,
+        was_requested: bool,
+        apply_chunks_done_callback: DoneApplyChunkCallback,
+    ) {
+        let hash = *block.hash();
+        let prev_hash = *block.header().prev_hash();
+        let _span = tracing::debug_span!(
+            target: "client",
+            "process_block",
```
Comment (matklad) [inline chain/client/src/client.rs:826]:
```suggestion
            "receive_block",
```

File: chain/client/src/client.rs [hunk 2]
```diff
@@ -807,6 +810,166 @@ impl Client {
...
+
+        match self.receive_block_impl(
+            block,
+            peer_id.clone(),
+            was_requested,
+            apply_chunks_done_callback,
+        ) {
+            Ok(_) => {}
+            // Log the errors here. Note that the real error handling logic is already
+            // done within process_block_impl, this is just for logging.
```
Comment (matklad) [inline chain/client/src/client.rs:843]: Let's but the comment on top of the whole match? Or maybe rewrite it as 

```
let res = self.receive_block_impl();
if let Err(err) = res {
            // Log the errors here. Note that the real error handling logic is already
            // done within process_block_impl, this is just for logging.
}
```

THe problem I am trying to solve here is that it's difficult at a glance to tell that we really just logging the errors here, without any kind of processing.

File: chain/client/src/client.rs [hunk 3]
```diff
@@ -807,6 +810,166 @@ impl Client {
...
+    /// Then it process the block header. If the header if valid, broadcast the block to its peers
+    /// Then it starts the block processing process to process the full block.
+    pub(crate) fn receive_block_impl(
+        &mut self,
+        block: Block,
+        peer_id: PeerId,
+        was_requested: bool,
+        apply_chunks_done_callback: DoneApplyChunkCallback,
+    ) -> Result<(), near_chain::Error> {
+        let prev_hash = *block.header().prev_hash();
```
Comment (matklad) [inline chain/client/src/client.rs:877]: move prev-hash calculation closer to the use?

File: chain/client/src/client_actor.rs [hunk 4]
```diff
@@ -1239,10 +1247,20 @@ impl ClientActor {
     fn produce_block(&mut self, next_height: BlockHeight) -> Result<(), Error> {
         let _span = tracing::debug_span!(target: "client", "produce_block", next_height).entered();
         if let Some(block) = self.client.produce_block(next_height)? {
-            let peer_id = self.node_id.clone();
             // We’ve produced the block so that counts as validated block.
```
Comment (matklad) [inline chain/client/src/client_actor.rs:1250]: misplaced commet

File: chain/client/src/client.rs [hunk 5]
```diff
@@ -807,6 +810,166 @@ impl Client {
...
+    }
+
+    /// To protect ourselves from spamming, we do some pre-check on block height before we do any
+    /// processing. This function returns true if the block height is valid.
+    fn check_block_height(
+        &self,
+        block: &Block,
+        was_requested: bool,
+    ) -> Result<bool, near_chain::Error> {
+        let head = self.chain.head()?;
```
[the remark this answers is not in the document]
Comment (mzhangmzz) [reply chain/client/src/client.rs:906]: Yep!

File: chain/client/src/client.rs (hunk 3 above)
Comment (mzhangmzz) [reply chain/client/src/client.rs:877]: `block` is moved later so prev_hash has to be calculated before that

**Pull request merged**